### PR TITLE
Issue/2425 load flash dynamically

### PR DESCRIPTION
**Pull request by e-carlin**
#2461 was closed inadvertently (maybe because the base branch was pruned). Reopening here.

### CODE REVIEW

**Comment by e-carlin** [comment]
@robnagler all comments addressed. Ready for another review.

File: sirepo/pkcli/admin.py [hunk 1]
```diff
@@ -77,9 +125,21 @@ def move_user_sims(target_uid=''):
         shutil.move(lib_file, target)
 
 
+def proprietary_sim_type_dir(sim_type):
+    return srdb.root().join(cfg.proprietary_code_dir, sim_type)
```
Comment (robnagler) [inline sirepo/pkcli/admin.py:129]: This doesn't need to be a cfg. It's "our" namespace so we can do what we want with it, and it isn't something we can change by just by changing the cfg. 

proprietary_sim_type_dir is proprietary_code_dir. No need for a different name.

File: sirepo/template/flash.py [hunk 2]
```diff
@@ -410,6 +408,17 @@ def _cell_size(f, refine_max):
...
 
+def _extract_zip(data, run_dir):
+    import os
+    import sirepo.pkcli.flash
+    import stat
+    import zipfile
+    zipfile.ZipFile(run_dir.join(_SIM_DATA.proprietary_lib_file_basename(data))).extractall()
+    # extractall() doesn't maintain file permissions
+    # https://bugs.python.org/issue15795
+    os.chmod(run_dir.join(sirepo.pkcli.flash.EXE_NAME), stat.S_IXUSR)
```
Comment (robnagler) [inline sirepo/template/flash.py:419]: I think this import should be the other way around. We don't want to import pkcli.flash here. pkcli.flash should get definitions from the template, or better yet, sim_data.

File: sirepo/template/flash.py [hunk 3]
```diff
@@ -424,10 +433,11 @@ def _cell_size(f, refine_max):
     ],
 }
 
+
 def _generate_parameters_file(data):
     res = ''
     names = {}
-    for line in pkio.read_text(_FLASH_UNITS_PATH[data.models.simulation.flashType]).split('\n'):
+    for line in pkio.read_text(_SETUP_UNITS_FILE).split('\n'):
```
Comment (robnagler) [inline sirepo/template/flash.py:440]: I think this should be in sim_data, too.

File: tests/pkcli/roles_test.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            sirepo.auth.role_for_sim_type(fc.sr_sim_type),
+        )
+
+    def _check_file(exists=True):
+        pkunit.pkeq(
+            [_PROPRIETARY_FILE] if exists else [],
+            [x.basename for x in pkio.walk_tree(fc.sr_user_dir(), _PROPRIETARY_FILE)],
+        )
+
+    _PROPRIETARY_FILE = 'CapLaser.zip'
```
Comment (robnagler) [inline tests/pkcli/roles_test.py:32]: this is not a global so it shouldn't be all caps.

**Comment by e-carlin** [comment]
@robnagler all comments addressed.

**Pull request merged**